src/confluent_kafka/aio/_AIOProducer.py

```diff
@@ -60,47 +142,254 @@ async def _call(self, blocking_task, *args, **kwargs):
+            topic_groups[topic]['callbacks'].append(msg_data.get('user_callback'))
+            topic_groups[topic]['indices'].append(i)
+        
+        # Clear buffers immediately (reduce memory pressure)
+        self._message_buffer.clear()
+        self._buffer_futures.clear()
+        
+        # Process each topic group separately
+        for topic, group_data in topic_groups.items():
+            # Skip if target_topic is specified and doesn't match
+            if target_topic is not None and topic != target_topic:
+                # Re-add messages for other topics back to buffer
+                for j, msg_data in enumerate(group_data['messages']):
+                    self._message_buffer.append(msg_data)
+                    self._buffer_futures.append(group_data['futures'][j])
+                continue
+            
+            # Convert messages to batch format
+            batch_messages = []
+            for msg_data in group_data['messages']:
+                batch_msg = {
+                    'value': msg_data['value'],
+                    'key': msg_data['key']
+                }
+                
+                # Add optional fields if present
+                if 'partition' in msg_data:
+                    batch_msg['partition'] = msg_data['partition']
+                if 'timestamp' in msg_data:
+                    batch_msg['timestamp'] = msg_data['timestamp']
+                if 'headers' in msg_data:
+                    batch_msg['headers'] = msg_data['headers']
+                
+                batch_messages.append(batch_msg)
+            
+            # Setup callback with pre-allocated state
+            batch_futures = group_data['futures']
+            batch_user_callbacks = group_data['callbacks']
+            callback_state = {'index': 0, 'futures': batch_futures, 'callbacks': batch_user_callbacks}
+            
+            def batch_delivery_callback(err, msg):
+                idx = callback_state['index']
+                if idx < len(callback_state['futures']):
+                    future = callback_state['futures'][idx]
+                    user_callback = callback_state['callbacks'][idx]
+                    
+                    if err:
+                        future.set_exception(_KafkaException(err))
+                    else:
+                        future.set_result(msg)
+                        if user_callback:
+                            try:
+                                user_callback(err, msg)
+                            except Exception:
+                                pass
+                    
+                    callback_state['index'] += 1
+            
+            try:
+                # Call produce_batch and poll for this topic
+                await self._call(self._produce_batch_and_poll, topic, batch_messages, batch_delivery_callback)
+                        
+            except Exception as e:
```

Comment:
Don't forget to re-raise the exception after marking as failed (and add a test for this case)